racingcar/racingcar.go

```diff
@@ -1,6 +1,52 @@
 package racingcar
 
-// 예시 함수, 구현 시 제거
-func Hello() string {
-	return "Hello, Racing Car!"
+import (
+	"math/rand"
+	"strings"
+	"time"
+)
+
+type Car struct {
+	Name     string
+	Position int
+}
+
+func (c *Car) Move() {
+	c.Position += 1
+}
+
+func (c Car) String() string {
+	return strings.Repeat("-", c.Position)
+}
+
+type CarGame struct {
+	Cars []*Car
+	rule Rule
+}
+
+func NewCarGame(cars []*Car, rule Rule) CarGame {
+	return CarGame{Cars: cars, rule: rule}
+}
+
+// Rule 인터페이스
+type Rule interface {
+	Check(n int) bool
+}
+
+// RandomRule은 Rule 인터페이스를 구현하는 구조체
+type RandomRule struct{}
+
+func (r RandomRule) Check(n int) bool {
+	rand.Seed(time.Now().UnixNano())
+	randomValue := rand.Intn(10)
+	return randomValue >= n
+}
```

Comment:
구조체 말고 함수를 그대로 넘기는건 어떨까요? Go는 함수 자체가 타입이 될 수 있어서 함수형 인터페이스처럼 구현할 필요가 없어요.

```suggestion
type Rule func(n int) bool

func RandomRule(n int) bool {
	rand.Seed(time.Now().UnixNano())
	randomValue := rand.Intn(10)
	return randomValue >= n
}
```

Reply:
오 그러면 시그니쳐만 같으면 위의 함수를 해당 타입으로 사용할 수 있는 건가요?

Reply:
넵. 더욱 정확하게 말하자면 타입을 지정할 필요도 없어요

```go
type CarGame struct {
	Cars []*Car
	rule func(int) bool
}

func NewCarGame(cars []*Car, rule func(int) bool) CarGame {
	return CarGame{Cars: cars, rule: rule}
}
```